Added tests for DiffableDataSource.

jaimeazevedo opened the pull request:
## What was done?

This PR contains some tests for DiffableDataSource functions.

AnthonyEgerton commented on Tests/DiffableDataSourceTests/DiffableTests.swift:
```diff
@@ -2,10 +2,104 @@ import XCTest
+        XCTAssertFalse(dataSource.tableView(tableView, canEditRowAt: IndexPath(row: 0, section: 2)))
+    }
+
+    func testSectionAt() {
+        XCTAssertNoThrow(try dataSource.replace(with: content, animatingDifferences: false))
```
You can just define the test method as `throws` and then it will catch any thrown errors and report them as test failures.
The call to `dataSource.replace` is more of an `Act` statement rather than an `Assert` statement in terms of the [Arrange, Act, Assert](http://wiki.c2.com/?ArrangeActAssert) pattern.
```suggestion
    func testSectionAt() throws {
        try dataSource.replace(with: content, animatingDifferences: false)
```

AnthonyEgerton commented on Tests/DiffableDataSourceTests/DiffableTests.swift:
```diff
@@ -2,10 +2,104 @@ import XCTest
+        }
+    }
+
+    func testRowAt() {
+        let content = [topSection, middleSection, bottomSection]
+
+        XCTAssertNoThrow(try dataSource.replace(with: content, animatingDifferences: false))
```
You can just define the test method as `throws` and then it will catch any thrown errors and report them as test failures.
The call to `dataSource.replace` is more of an `Act` statement rather than an `Assert` statement in terms of the [Arrange, Act, Assert](http://wiki.c2.com/?ArrangeActAssert) pattern.
```suggestion
    func testRowAt() throws {
        let content = [topSection, middleSection, bottomSection]

        try dataSource.replace(with: content, animatingDifferences: false)
```

AnthonyEgerton commented on Tests/DiffableDataSourceTests/DiffableTests.swift:
```diff
@@ -2,10 +2,104 @@ import XCTest
+        XCTAssertNoThrow(try dataSource.replace(with: content, animatingDifferences: false))
+
+        for index in 0..<content.count {
+            XCTAssertTrue(dataSource.section(at: index) == content[index].identifier)
```
It's better to use XCTAssertEqual when you can as it will give more information if that assertion fails. 
Also when asserting with multiple values it's a good practice to provide more context in the error message as to which assertion failed. This can be very useful when trying to determine what failed during CI runs.
```suggestion
            XCTAssertEqual(
                dataSource.section(at: index),
                content[index].identifier,
                "Section wrong at index(\(index))"
            )
```

AnthonyEgerton commented on Tests/DiffableDataSourceTests/DiffableTests.swift:
```diff
@@ -2,10 +2,104 @@ import XCTest
+
+        for section in 0..<content.count {
+            for row in 0..<content[section].rows.count {
+                XCTAssertTrue(dataSource.row(at: IndexPath(row: row, section: section)) == content[section].rows[row])
```
It's better to use XCTAssertEqual when you can as it will give more information if that assertion fails. 
Also when asserting with multiple values it's a good practice to provide more context in the error message as to which assertion failed. This can be very useful when trying to determine what failed during CI runs.
```suggestion
                let indexPath = IndexPath(row: row, section: section)
                XCTAssertEqual(
                    dataSource.row(at: indexPath),
                    content[section].rows[row],
                    "Row wrong at indexPath:(\(indexPath))"
                )
```

jaimeazevedo commented:
Thank you for the review, I've learned some stuff 👍 I addressed your comments and also removed the content property as it does not make much sense in the AAA pattern.

NSMyself reviewed (approved):
I think it'd be nice if in the future we could replace throwing an error with returning a Result instead but that is besides the point of this PR.

Good work 👍
